### PR TITLE
Collect metrics once per lambda invocation

### LINKED ISSUE #1534: [META 637] Enable custom metrics collection with AWS Lambda
A few users have expressed interest in custom metrics (whether through our metrics system or prometheus) in lambda. We should be able to do a single, synchronous collection of the metrics before flush, but it should be config-gated and not include any of the base metrics, just custom/prometheus

See meta issue and spec for the description and details:
- Meta issue: https://github.com/elastic/apm/issues/637
- Spec issue: https://github.com/elastic/apm/issues/638

### PR DESCRIPTION
**Pull request by basepi**
## What does this pull request do?

1. Collects metrics once per lambda invocation. Disables breakdown and CPU metrics, so only custom metrics and prometheus metrics will be collected.
2. Allows MetricSet classes to be added to the MetricsRegistry. This allows inline metrics to be created, rather than relying on importing based on the class path.
3. Renames MetricsSet to MetricSet for consistency.

Remaining:

- [x] Add docs for custom metrics
- [x] Manual testing for AWS lambda
- [x] Write some unit tests
- [x] Rename `_metrics` to `metrics` so we're not accessing a private attribute?

## Related issues

Closes #1534

### CODE REVIEW

**Comment by basepi** [comment]
I still need to do some testing around AWS lambda use, but I added a unit test and some docs. The docs in particular could use some review.

File: docs/metrics.asciidoc [hunk 1]
```diff
@@ -160,3 +160,56 @@ All metrics collected from `prometheus_client` are prefixed with `"prometheus.me
...
 ===== Beta limitations
  * The metrics format may change without backwards compatibility in future releases.
+
+[float]
+[[custom-metrics]]
+=== Custom Metrics
+
+Custom metrics allow you to send your own metrics to Elasticsearch.
+
+The most commond way to send custom metrics is with the Prometheus metrics set.
```
Comment (bmorelli25) [inline docs/metrics.asciidoc:170]:
```suggestion
The most common way to send custom metrics is with the Prometheus metrics set.
```

File: docs/metrics.asciidoc [hunk 2]
```diff
@@ -160,3 +160,56 @@ All metrics collected from `prometheus_client` are prefixed with `"prometheus.me
 [[prometheus-metricset-beta]]
 ===== Beta limitations
  * The metrics format may change without backwards compatibility in future releases.
+
+[float]
+[[custom-metrics]]
+=== Custom Metrics
```
Comment (bmorelli25) [inline docs/metrics.asciidoc:166]: There's a jump table on L11-L13. We should add en entry that links to this new topic.

File: docs/metrics.asciidoc [hunk 3]
```diff
@@ -160,3 +160,56 @@ All metrics collected from `prometheus_client` are prefixed with `"prometheus.me
...
+
+Custom metrics allow you to send your own metrics to Elasticsearch.
+
+The most commond way to send custom metrics is with the Prometheus metrics set.
+See <<prometheus-metricset>> for more information.
+
+However, you can also use your own metrics set. If you collect the metrics
+manually in your code, you can use the base `MetricsSet` class:
```
Comment (bmorelli25) [inline docs/metrics.asciidoc:174]: Maybe remove the line break on L172 and turn this into one paragraph?

File: docs/metrics.asciidoc [hunk 4]
```diff
@@ -160,3 +160,56 @@ All metrics collected from `prometheus_client` are prefixed with `"prometheus.me
...
  * The metrics format may change without backwards compatibility in future releases.
+
+[float]
+[[custom-metrics]]
+=== Custom Metrics
+
+Custom metrics allow you to send your own metrics to Elasticsearch.
+
+The most commond way to send custom metrics is with the Prometheus metrics set.
+See <<prometheus-metricset>> for more information.
```
Comment (bmorelli25) [inline docs/metrics.asciidoc:171]: I would just turn the `Prometheus metrics set` text in the previous sentence into a link and omit this sentence.

File: docs/metrics.asciidoc [hunk 5]
```diff
@@ -160,3 +160,56 @@ All metrics collected from `prometheus_client` are prefixed with `"prometheus.me
...
+method, where you can gather and set metrics before they are collected and sent
+to Elasticsearch.
+
+You can instantiate and add your `MetricsSet` instance as shown in the example above,
+or you can pass in an import string for the class:
+
+[source,python]
+----
+client = elasticapm.Client()
+client._metrics.register("myapp.metrics.MyAwesomeMetricsSet")
```
Comment (beniwohli) [inline docs/metrics.asciidoc:199]: I'd rather document the `metrics_sets` config option to handle this case, tbh

Comment (basepi) [reply docs/metrics.asciidoc:199]: Good call, I was too deep in the code and totally missed/forgot that that was the typical way to add custom metrics sets.

File: docs/metrics.asciidoc [hunk 6]
```diff
@@ -160,3 +160,56 @@ All metrics collected from `prometheus_client` are prefixed with `"prometheus.me
...
+However, you can also use your own metrics set. If you collect the metrics
+manually in your code, you can use the base `MetricsSet` class:
+
+[source,python]
+----
+from elasticapm.metrics.base_metrics import MetricsSet
+
+client = elasticapm.Client()
+metrics_set = MetricsSet(client._metrics)
+client._metrics.register("my_awesome_metrics", metrics_set)
```
[the remark this answers is not in the document]
Comment (basepi) [reply docs/metrics.asciidoc:182]: Excellent idea. Much more consistent and simple. Thanks!

**Comment by basepi** [comment]
Ah, lint is failing as referenced in #1658

File: elasticapm/metrics/base_metrics.py [hunk 7]
```diff
@@ -56,25 +57,36 @@ def __init__(self, client, tags=None):
...
+
+        :param metricset: a string with the import path of the metricset class,
+            or a class object that can be used to instantiate the metricset.
+            If a class object is used, you can use the class object or
+            `metricset.__name__` to retrieve the metricset using `get_metricset`.
+        :return: the metricset instance
         """
-        if class_path in self._metricsets:
-            return
+        class_id = metricset if isinstance(metricset, str) else metricset.__name__
```
Comment (beniwohli) [inline elasticapm/metrics/base_metrics.py:70]: using only `metricset.__name__` here could lead to double registrations here I think. We should use the full class path instead, i.e.: 

```suggestion
        class_id = metricset if isinstance(metricset, str) else f"{metricset.__module__}.{metricset.__name__}"

File: elasticapm/metrics/base_metrics.py [hunk 8]
```diff
@@ -56,25 +57,36 @@ def __init__(self, client, tags=None):
...
+                    class_obj = import_string(metricset)
+                    self._metricsets[metricset] = class_obj(self)
+                except ImportError as e:
+                    logger.warning("Could not register %s metricset: %s", metricset, str(e))
+            else:
+                self._metricsets[class_id] = metricset(self)
+        return self._metricsets.get(class_id)
+
+    def get_metricset(self, metricset: Union[str, type]) -> "MetricSet":
+        metricset = metricset if isinstance(metricset, str) else metricset.__name__
```
Comment (beniwohli) [inline elasticapm/metrics/base_metrics.py:85]:
```suggestion
        metricset = metricset if isinstance(metricset, str) else f"{metricset.__module__}.{metricset.__name__}"

**Pull request merged**